rust/lanzatool/src/install.rs

```diff
@@ -8,37 +9,71 @@ use nix::unistd::sync;
 use tempfile::tempdir;
 
 use crate::esp::EspPaths;
-use crate::generation::Generation;
+use crate::gc::Roots;
+use crate::generation::{Generation, GenerationLink};
 use crate::pe;
 use crate::signature::KeyPair;
 
 pub struct Installer {
+    gc_roots: RefCell<Roots>,
     lanzaboote_stub: PathBuf,
     key_pair: KeyPair,
+    configuration_limit: u64,
     esp: PathBuf,
-    generations: Vec<PathBuf>,
+    generation_links: Vec<PathBuf>,
 }
 
 impl Installer {
     pub fn new(
         lanzaboote_stub: PathBuf,
         key_pair: KeyPair,
+        configuration_limit: u64,
         esp: PathBuf,
-        generations: Vec<PathBuf>,
+        generation_links: Vec<PathBuf>,
     ) -> Self {
         Self {
+            gc_roots: RefCell::new(Roots::new()),
             lanzaboote_stub,
             key_pair,
+            configuration_limit,
             esp,
-            generations,
+            generation_links,
         }
     }
 
     pub fn install(&self) -> Result<()> {
-        for toplevel in &self.generations {
-            let generation_result = Generation::from_toplevel(toplevel)
-                .with_context(|| format!("Failed to build generation from toplevel: {toplevel:?}"));
+        let mut links = self
+            .generation_links
+            .iter()
+            .map(GenerationLink::from_path)
+            .collect::<Result<Vec<GenerationLink>>>()?;
+
+        // A configuration limit of 0 means there is no limit
+        if self.configuration_limit > 0 {
+            // Sort the links by version
+            links.sort();
```

Comment:
You want `.sort_by_key(|l| l.version)` instead of implementing `PartialEq`, `Eq` and `Ord` for `GenerationLink`.

The problem with your trait implementations is that they do not actually implement the traits as they are intended. Two `GenerationLink` instances will compare as equal even though they have different paths. This may be surprising in other circumstances.

Reply:
This is what I was looking for! I guess I was just overcome with small brain syndrome and didn't bother to look into a more elegant solution. Thanks for the pointer!